docs/engineering/architecture/models.md

```diff
@@ -0,0 +1,215 @@
+| date\_joined             | timestamp with time zone |                  |
+| full\_name               | character varying(255)   |                  |
+| short\_name              | character varying(255)   |                  |
+| password\_change\_needed | boolean                  |                  |
+
+## DBServer
+
+| Column      | Type                     | Notes    |
+|-------------|--------------------------|----------|
+| id          | integer                  | pkey     |
+| created\_at | timestamp with time zone |          |
+| updated\_at | timestamp with time zone |          |
+| host        | character varying        | not null |
+| port        | integer                  | not null |
+
+`(host, port)` pair is unique. 
+
+Theoretically, we could also split the host out, but that seems like premature optimization.
+
+We could consider making the `host` and `port` nullable when we're supporting `.pgpass`.
+
+## Database
+
+| Column                          | Type                     | Notes                                       |
+|---------------------------------|--------------------------|---------------------------------------------|
+| id                              | integer                  | pkey                                        |
+| created\_at                     | timestamp with time zone |                                             |
+| updated\_at                     | timestamp with time zone |                                             |
+| db\_name                        | text                     | not null                                    |
+| display\_name                   | text                     | not null; unique                            |
+| db\_server                      | integer                  | not null; references DBServer(id)           |
+| editable                        | boolean                  |                                             |
+| default\_db\_server\_credential | integer                  | not null; references DBServerCredential(id) |
+
+`(db_server, db_name)` is unique. We could consider making `db_name` nullable when supporting `.pgpass`. If a Mathesar Admin user doesn't have an entry in `UserDBRoleMap` for a given database, they will use the `default_credential` defined here to connect.
+
+## DBServerCredential
+
+| Column      | Type                     | Notes                             |
+|-------------|--------------------------|-----------------------------------|
+| id          | integer                  | pkey                              |
+| created\_at | timestamp with time zone |                                   |
+| updated\_at | timestamp with time zone |                                   |
+| username    | character varying        | not null                          |
+| password    | character varying        | encrypted; not null               |
+| db\_server  | integer                  | not null; references DBServer(id) |
```

Comment:
I wasn't planning for it to be unique. As you mention, we may have reason to have more than copy of such a pair.

Reply:
> configure the credentials when we could allow them to select from an already saved credential

Assuming that credentials are currently configured within the database page on the UX, this would break the scope of what it means to save a new connection since it'll break the boundaries of a database on the UI. It would also be a security concern since we don't want credentials configured within one database to be leaked to another.

>  I'd imagine we'd actually want to validate that the newly-entered password agrees with the stored password

It'll introduce new complexities such as figuring out which is the correct password. The stored password might be outdated. If we were to re-validate, do we now also update the old one? All these scenarios would complicate UX.

> To me, these are UX considerations that I'd be inclined to make based on the models
> I'd want to approach the models from a perspective of data integrity and normalization

I think the usecase should determine the modal structure and not the other way around. The purpose of this model is also to support the UX. Keeing it flexible would help us experiment with the UX more than having to restrict our UX affecting usability for little functional benefit.

I do think we can talk about this later, and merging the PR in.

Reply:
@mathemancer, @pavish, and I discussed this on a call today. On the call I suggested two different directions to go, which I outline below as (B) and (C).

(In these diagrams I'm using ⭐ to indicate multi-column unique constraints.)

- **(A)** This is what the PR currently has (listed here for reference)

    ![Screenshot_20240329_130345](https://github.com/mathesar-foundation/mathesar-wiki/assets/42411/09bc3238-0eb6-459f-8cb0-416ee6909f0f)


- **(B)** Databases and credentials are siblings within a DB server.

    ![image](https://github.com/mathesar-foundation/mathesar-wiki/assets/42411/793e1248-0c82-41e3-a625-6de51b0bcaea)

    
    - Credentials would be shared across separate databases.
    - Updating a password in one place would apply to multiple databases.
    - The same is true for port numbers. To edit a port number, we'd need to make it clear to users That they'd be editing the port number of the `DBServer` object which could apply to many databases.
    - Databases, servers, and credentials could all be maintained separately from one another.
    - It would be possible to keep a credential stored within Mathesar while deleting all databases. Then you could connect to a new database using that same credential.

- **(C)** Each database owns a set of credentials.

    ![image](https://github.com/mathesar-foundation/mathesar-wiki/assets/42411/b7a40430-25f7-4fcf-8d50-0c68f9d38b3f)

    - Credentials and servers are both scoped _underneath_ databases.
    - Since a database only has one server, the sever details are inlined into the database model.
    - Separate databases require separate credentials, and those credentials must be duplicated.
    - Any changes you make to server or credential settings within a database will only apply to that database.
    - Deleting a database from Mathesar would delete all the credentials associated with it.

The point I'm trying to make is that (B) and (C) are both sound approaches, while (A) is error-prone and hacky (in my opinion). With (A) the DB structure does not provide clarity on whether a credential should be shared or duplicated across multiple databases. That's the sort of ambiguity that I strive to minimize when designing database schemas. Ideally any database state should be valid application state. So if our application expects separate sets of credentials per-database, then we should enforce that logic at the data layer to avoid bugs. If we're _not sure_ if we want separate credential sets or not, then we should figure that out now instead of allowing for both structures to coexist in the data layer.

From a product perspective, I think it's worth considering (B) and (C) — and I think we ought to do this _before_ getting into anything with Figma (for the sake of time). To me, (C) seems simplest which is a strong point in its favor. Design and implementation would be much more straightforward than with (C). Approach (B) seems more powerful. But perhaps power users would be better served by additional features (post-beta) which allow for configuration code (e.g. in yaml). Honestly I don't think I have a strong opinion on (B) vs (C). I just would prefer that we avoid (A).

Apologies that I failed to notice the lack of unique constraint. I would have mentioned it sooner had I noticed.

Reply:
Another product-level thought on (B) vs (C)...

If a team of people is using Mathesar as an Airtable replacement, we need to give them something equivalent to an Airtable "Base" ­— i.e. a self-contained project that they create from scratch within Mathesar.

- If we want them to use _schemas_ to separate their projects, then (C) works fine. However we run into challenges with schema level permissions.
- If we want them to use _databases_ to separate their projects, then (B) seems slightly superior because the process of creating and removing databases could be more easily decoupled from managing credentials. Creating a new Airtable base is quite easy. Requiring users to enter a Postgres password for that step would be a usability impediment.

Reply:
BTW, Brent wanted the code for these diagrams. It's here:

https://dbdiagram.io/d/Mathesar-internal-AB-6604b63cae072629ce23f20a

https://dbdiagram.io/d/Mathesar-internal-C-6606dd2e37b7e33fd70f376a

Reply:
Taking options (B) and (C) as a given, I strongly prefer (B), since it's more flexible and more representative of the actual underlying structure of the objects involved. The only reason I still support (A) is that we already have a UI that needs that flexibility (or ambiguity), and (A) can easily be modified to (B) if we go that route.

I'm suspicious of (C), and any UI that depends on (C), since it introduces a misconception that we (and the user) will have to maintain, i.e., that a role on the PostgreSQL server is namespaced under a Database on that server. We have typically regretted creating simplifying abstractions that misrepresent the way PostgreSQL actually works on this project.

Regarding the concerns of @pavish w.r.t. organizational multitenancy, I think it's perfectly reasonable for that to be on a Mathesar installation level for now (rather than on a Database level).

Reply:
Cool. And if you both want to go with (A) I'll stop fussing about it, though I might have some more pointed questions and critique once we have designs in Figma.

Reply:
This discussion seems like bikeshedding to me at this point, we can always add a unique constraint later and I don't see why we need to make a decision about whether or how to share credentials between databases right now. I'm going to make the executive decision that we're going with (A).

If the UX we end up with means that (A) can end up with invalid DB states, then we can implement (B). It's not like backend changes are verboten after this gets merged.

I do not think we should do (C) in any situation because it's not accurate to how credentials actually work and trying to build on an incorrect abstraction will probably cause us problems later.

Reply:
If we were to make a decision about whether to share credentials now, I'd still argue for (A) anyway because implementing sharing credentials well seems like it would add scope to the beta (extra UX flows, security considerations, etc.) and I'm not sure that the feature of sharing credentials is valuable enough to add scope.

Reply:
As already stated, I'm in agreement with @mathemancer and @kgodey to go with (A).